### DISCUSSION TITLE
Better way to get capture groups from a regex?

### DISCUSSION THREAD

**Discussion by fundef1** [duckdb/duckdb#6945] [Q&A] [answered]
Hi, 
i'm dealing with a convoluted regex where I need retrieve all (non-empty) capture groups.

the best I came up with is calling the regex over and over and retrieving the capture groups one by one with regexp_extract.
```
regexp_extract(line,'{{regex}}',1) as cg1,
regexp_extract(line,'{{regex}}',2) as cg2,
regexp_extract(line,'{{regex}}',3) as cg3,
regexp_extract(line,'{{regex}}',4) as cg4,
regexp_extract(line,'{{regex}}',5) as cg5,
regexp_extract(line,'{{regex}}',6) as cg6,
regexp_extract(line,'{{regex}}',7) as cg7,
regexp_extract(line,'{{regex}}',8) as cg8,
regexp_extract(line,'{{regex}}',9) as cg9
```
which is neither regexp_extract's intended use (I think)
nor hardly efficient (I wouldn't think this gets optimized into a single call into RE2)
and more importantly there's an (undocumented) limit on 9 capture groups. :-(

in Postgres I'd use REGEXP_MATCH  (note: not REGEXP_MATCH**ES**)
and work with the returned array.

any ideas on how to achieve similar functionality with DuckDB?
Thanks!

PS: yes, such string manipulation/extraction is probably best done outside the database, but  "reasons"

**Comment by Alex-Monahan** [comment, marked as answer]
Howdy! Have a look at the great work done by @hawkfish in this PR to see if that would meet your use case!
https://github.com/duckdb/duckdb/pull/6918

**Comment by hawkfish** [comment]
Thanks @Alex-Monahan 

@fundef1 I set up a discussion in #6882 and basically it does what you are asking, but creates a struct instead of an array.

**Comment by Alex-Monahan** [comment]
As a note, the PR above plus the `my_struct.*` syntax likely would do the trick!

**Reply by fundef1** [reply]
Great! struct or array doesn't matter really.
and since this in the pipeline... any chance of supporting named capture groups?

**Reply by hawkfish** [reply]
I thought about that but this was the simplest approach. Main thing would be the function call sequence, but we could always just come up with another function.

**Reply by fundef1** [reply]
okay thanks, a struct will work so I'll wait for that functionality to be in a release.

Being able to refer to capture groups by name would simplify the maintenance of capture groups of the regex vs columns in the database. 
But I'm happy with the struct solution, can't wait for the next release ;-)

**Comment by Mause** [comment]
> Hi, i'm dealing with a convoluted regex where I need retrieve all (non-empty) capture groups.
> 
> the best I came up with is calling the regex over and over and retrieving the capture groups one by one with regexp_extract.
> 
> ```
> regexp_extract(line,'{{regex}}',1) as cg1,
> regexp_extract(line,'{{regex}}',2) as cg2,
> regexp_extract(line,'{{regex}}',3) as cg3,
> regexp_extract(line,'{{regex}}',4) as cg4,
> regexp_extract(line,'{{regex}}',5) as cg5,
> regexp_extract(line,'{{regex}}',6) as cg6,
> regexp_extract(line,'{{regex}}',7) as cg7,
> regexp_extract(line,'{{regex}}',8) as cg8,
> regexp_extract(line,'{{regex}}',9) as cg9
> ```
> 
> which is neither regexp_extract's intended use (I think) nor hardly efficient (I wouldn't think this gets optimized into a single call into RE2) and more importantly there's an (undocumented) limit on 9 capture groups. :-(
> 
> in Postgres I'd use REGEXP_MATCH (note: not REGEXP_MATCH**ES**) and work with the returned array.
> 
> any ideas on how to achieve similar functionality with DuckDB? Thanks!
> 
> PS: yes, such string manipulation/extraction is probably best done outside the database, but "reasons"

You probably want `regexp_extract_all`, right? See https://github.com/duckdb/duckdb/pull/6685

**Reply by Tishj** [reply]
No not quite, extract_all still takes a single group.
The difference there is that it gets all matches for the given group, not just the first match

**Reply by fundef1** [reply]
Indeed, I'm looking for (the first match) of **all** groups, not all matches of a **single** group.

**Comment by hawkfish** [comment]
The struct version was merged today.

**Reply by fundef1** [reply]
When would this be available in a release?

(or if someone could provide a linux binary) 
I'd be happy to take if for a test ride.

**Reply by hawkfish** [reply]
Our next release is probably early May.  The Linux nightly releases are [here](https://github.com/duckdb/duckdb/actions?query=branch%3Amaster+event%3Arepository_dispatch+workflow%3ALinuxRelease)

**Comment by hawkfish** [comment]
I'm now wondering if the way to get the named capture groups is with an empty list? That is currently an error.

**Reply by fundef1** [reply]
An empty list would work.  Although I wonder whether the list is needed at all...
not using a list would bring it in line with  (py)arrow's extract_regex function.
(renaming/aliasing) might also avoid some confusion with extract_all, see above

```
pyarrow.compute.extract_regex(strings, /, pattern, *, options=None, memory_pool=None)
Extract substrings captured by a regex pattern.

For each string in strings, match the regular expression and, if successful, emit a struct with field names and values coming from the regular expression’s named capture groups. If the input is null or the regular expression fails matching, a null output value is emitted.
```
https://arrow.apache.org/docs/python/generated/pyarrow.compute.extract_regex.html#pyarrow.compute.extract_regex

just my 2 cents, an empty list is fine.

**Reply by hawkfish** [reply]
Unfortunately, the two argument version extracts capture group 0 (i.e., the entire match).

**Reply by fundef1** [reply]
yes, that's why I suggested a new "extract_regex" function instead.

**Comment by fundef1** [comment]
(sorry for waiting so long with picking this up again) - I'm now on 8.0.1, and I found the behaviour for named regexes confusing/unexpected wrt to the use/naming as a struct: i.e. it seems the capture groups are not referenced by name, but by their position in the struct. that's probably  bit cryptic... 

from the docs: 
``` regexp_extract('2023-04-15 07:59:56', '^(\d+)-(\d+)-(\d+) (\d+):(\d+):(\d+)', ['y', 'm', 'd']) -- {'y':'2023', 'm':'04', 'd':'15'} ```
modified with named groups:
```select regexp_extract('2023-04-15 07:59:56', '^(?P<y>\d+)-(?P<m>\d+)-(?P<d>\d+) (?P<h>\d+):(?P<mm>\d+):(?P<s>\d+)', ['y', 'm', 'd','hh','mm','s'])  -- {'y': 2023, 'm': 04, 'd': 15} ```

but this is unexpected:
``` regexp_extract('2023-04-15 07:59:56', '^(?P<y>\d+)-(?P<m>\d+)-(?P<d>\d+) (?P<h>\d+):(?P<mm>\d+):(?P<s>\d+)', ['d', 'm', 'y']) -- {'d': 2023, 'm': 04, 'y': 15} ```

is this behaviour intentional?  (i'm guessing named groups are just ignored) - if it is intentional, any chance of seeing the 'empty list' approach in a release soon?

thanks!

**Reply by Tishj** [reply]
I agree this is weird behavior, and I think we should add support for named groups in this function.

For the second example you shared I get a different result, so I'm a little confused?
```sql
select regexp_extract('2023-04-15 07:59:56', '^(?P<y>\d+)-(?P<m>\d+)-(?P<d>\d+) (?P<h>\d+):(?P<mm>\d+):(?P<s>\d+)', ['y', 'm', 'd','hh','mm','s'])
----
{'y': 2023, 'm': 04, 'd': 15, 'hh': 07, 'mm': 59, 's': 56}
```

Also please try to use codeblocks, these are currently very hard to read / digest.

I think a good behavior here would be:
- If there are any named groups, output only the groups that have a (case-insensitive ?) match in `names`
- If there are no named groups, output all of the groups in order aliased by `names`

What you're proposing is to be able to provide an empty `names` list and use the names of the groups instead?
In that case we should probably verify that all groups are named, and throw otherwise

**Reply by fundef1** [reply]
first off, yes, second example is incorrect - that's a copy/paste error - reran it and this is the correct results
`{'y': 2023, 'm': 04, 'd': 15, 'hh': 07, 'mm': 59, 's': 56}`   (sorry, so used to using triple back-ticks hope this is more readable for you)

The empty list was a suggestion from @hawkfish 
>> I'm now wondering if the way to get the named capture groups is with an empty list? That is currently an error.
I suggested bringing the behaviour more inline with arrow's extract_regex but either implementation is fine. - see earlier comments.

If going the empty list route, my preference would be to not throw an error and return the named capture groups by name and the non-named groups by their number. Named capture groups have to start with an alphanumeric character so i guess this should be safe.
e.g (?P<name>....)  as struct.name and (...) as struct.1 

this might lead to some duplicates though, a quick check on the re2 syntax suggest that named capture groups are named and numbered: 

<html>
<body>


(re) | numbered capturing group (submatch)
-- | --
(?P&lt;name&gt;re) | named & numbered capturing group (submatch)


</body>
</html>
